### ISSUE TITLE
Feature request: controlling tracking from inside the program

### ISSUE THREAD

**Issue by CAD97** [rust-lang/miri#2353]
Specifically,

```rust
#[cfg(miri)]
extern "Rust" {
    /// Start tracking the pointee alloc id as if -Zmiri-track-alloc-id was used
    fn miri_start_tracking_alloc(ptr: *mut ());
    /// Stop tracking the pointee alloc id
    fn miri_stop_tracking_alloc(ptr: *mut ());
    /// Print the alloc id as if tracking was triggered
    fn miri_track_alloc(ptr: *mut ());
    
    /// Start tracking the pointer tag as if -Zmiri-track-pointer-tag was used
    fn miri_start_tracking_tag(ptr: *mut ());
    /// Stop tracking the pointer tag
    fn miri_stop_tracking_tag(ptr: *mut ());
    /// Print the pointer tag as if tracking was triggered
    fn miri_track_tag(ptr: *mut ());
}
```

Names are bikeshedable of course.

**Comment by RalfJung** [comment]
I agree something like this makes sense. However I think the details are not entirely obvious.
- Part of the point of tracking an alloc ID / ptr tag is to figure out when it got *created*. I that sense, these runtime APIs can never be the same kind of tracking as what the flags provide. `miri_start_tracking_alloc` would literally just be `miri_tell_me_when_this_gets_deallocated`.
- "Print the alloc id as if tracking was triggered" sounds like you want a `miri_get_alloc_id(ptr: *mut ()) -> u64`? This is certainly not "tracking" anything in any sense of the word "tracking" that I am aware of. ;)
- For pointer tags, the suggested types would be it impossible or at least very hard to track a reference. Casting it to a raw ptr will change its tag! So I think we need to make these functions take a `ptr: T` with a comment saying `ptr` must be a reference or raw pointer. But even then you won't be able to do this on a mutable reference since you would lose the reference (or reborrow it, which changes the tag). So maybe `ref_to_ptr: &T`, and the relevant tag is that of the pointer that the reference points to?
